sys/ztimer/init.c

```diff
@@ -266,15 +267,64 @@ void ztimer_init(void)
 #  else
     LOG_DEBUG("ztimer_init(): ZTIMER_USEC without conversion\n");
 #  endif
-#  ifdef CONFIG_ZTIMER_USEC_ADJUST_SET
-    LOG_DEBUG("ztimer_init(): ZTIMER_USEC setting adjust_set value to %i\n",
-              CONFIG_ZTIMER_USEC_ADJUST_SET );
-    ZTIMER_USEC->adjust_set = CONFIG_ZTIMER_USEC_ADJUST_SET;
-#  endif
-#  ifdef CONFIG_ZTIMER_USEC_ADJUST_SLEEP
-    LOG_DEBUG("ztimer_init(): ZTIMER_USEC setting adjust_sleep value to %i\n",
-              CONFIG_ZTIMER_USEC_ADJUST_SLEEP );
-    ZTIMER_USEC->adjust_sleep = CONFIG_ZTIMER_USEC_ADJUST_SLEEP;
+
+    /* warm-up time if set and needed */
+    if (IS_USED(MODULE_ZTIMER_AUTO_ADJUST) &&
+        !(IS_ACTIVE(CONFIG_ZTIMER_USEC_ADJUST_SET) &&
+          IS_ACTIVE(CONFIG_ZTIMER_USEC_ADJUST_SLEEP))) {
+        if (IS_ACTIVE(CONFIG_ZTIMER_AUTO_ADJUST_SETTLE)) {
+            ztimer_sleep(ZTIMER_USEC, CONFIG_ZTIMER_AUTO_ADJUST_SETTLE);
+        }
+    }
+
+    /* calculate or set 'adjust_set' */
+    if (IS_ACTIVE(CONFIG_ZTIMER_USEC_ADJUST_SET)) {
+        ZTIMER_USEC->adjust_set = CONFIG_ZTIMER_USEC_ADJUST_SET;
+    }
+    else if (IS_USED(MODULE_ZTIMER_AUTO_ADJUST)) {
+        int32_t overhead = ztimer_overhead_set(ZTIMER_USEC, 1000);
+        if (overhead > 0) {
+            ZTIMER_USEC->adjust_set = overhead + 1;
+        }
+        for (unsigned i = 0; i < 100; i++) {
+            int32_t overhead = ztimer_overhead_set(ZTIMER_USEC, 1000);
+            if (overhead < 0) {
+                ZTIMER_USEC->adjust_set += overhead;
+            }
+        }
+    }
+    if (ZTIMER_USEC->adjust_set) {
+        LOG_DEBUG("ztimer_init(): ZTIMER_USEC setting adjust_set value to %i\n",
+                  ZTIMER_USEC->adjust_set);
+    }
+
+    /* calculate or set 'adjust_sleep' */
+    if (IS_ACTIVE(CONFIG_ZTIMER_USEC_ADJUST_SLEEP)) {
+        ZTIMER_USEC->adjust_sleep = CONFIG_ZTIMER_USEC_ADJUST_SLEEP;
+    }
+    else if (IS_USED(MODULE_ZTIMER_AUTO_ADJUST)) {}
```

Comment:
These braces don't look like something you intended

Reply:
Thanks!